crates/node-builder/src/rpc.rs

```diff
@@ -0,0 +1,301 @@
+    fn on_rpc_started(
+        &self,
+        ctx: RpcContext<'_, Node>,
+        handles: RethRpcServerHandles,
+    ) -> eyre::Result<()>;
+}
+
+impl<Node, F> OnRpcStarted<Node> for F
+where
+    F: Fn(RpcContext<'_, Node>, RethRpcServerHandles) -> eyre::Result<()>,
+    Node: FullNodeComponents,
+{
+    fn on_rpc_started(
+        &self,
+        ctx: RpcContext<'_, Node>,
+        handles: RethRpcServerHandles,
+    ) -> eyre::Result<()> {
+        self(ctx, handles)
+    }
+}
+
+impl<Node: FullNodeComponents> OnRpcStarted<Node> for () {
+    fn on_rpc_started(&self, _: RpcContext<'_, Node>, _: RethRpcServerHandles) -> eyre::Result<()> {
+        Ok(())
+    }
+}
+
+/// Event hook that is called when the rpc server is started.
+pub trait ExtendRpcModules<Node: FullNodeComponents> {
+    /// The hook that is called once the rpc server is started.
+    fn extend_rpc_modules(&self, ctx: RpcContext<'_, Node>) -> eyre::Result<()>;
+}
+
+impl<Node, F> ExtendRpcModules<Node> for F
+where
+    F: Fn(RpcContext<'_, Node>) -> eyre::Result<()>,
+    Node: FullNodeComponents,
+{
+    fn extend_rpc_modules(&self, ctx: RpcContext<'_, Node>) -> eyre::Result<()> {
+        self(ctx)
+    }
+}
+
+impl<Node: FullNodeComponents> ExtendRpcModules<Node> for () {
+    fn extend_rpc_modules(&self, _: RpcContext<'_, Node>) -> eyre::Result<()> {
+        Ok(())
+    }
+}
+
+/// Helper wrapper type to encapsulate the [RethModuleRegistry] over components trait.
+#[derive(Debug)]
+pub struct RpcRegistry<Node: FullNodeComponents> {
+    pub(crate) registry: RethModuleRegistry<
+        Node::Provider,
+        Node::Pool,
+        NetworkHandle,
+        TaskExecutor,
+        Node::Provider,
+        Node::Evm,
+    >,
+}
+
+impl<Node: FullNodeComponents> Deref for RpcRegistry<Node> {
+    type Target = RethModuleRegistry<
+        Node::Provider,
+        Node::Pool,
+        NetworkHandle,
+        TaskExecutor,
+        Node::Provider,
+        Node::Evm,
+    >;
+
+    fn deref(&self) -> &Self::Target {
+        &self.registry
+    }
+}
+
+impl<Node: FullNodeComponents> DerefMut for RpcRegistry<Node> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.registry
+    }
+}
+
+impl<Node: FullNodeComponents> Clone for RpcRegistry<Node> {
+    fn clone(&self) -> Self {
+        Self { registry: self.registry.clone() }
+    }
+}
+
+/// Helper container to encapsulate [RethModuleRegistry], [TransportRpcModules] and [AuthRpcModule].
+///
+/// This can be used to access installed modules, or create commonly used handlers like
+/// [reth_rpc::EthApi], and ultimately merge additional rpc handler into the configured transport
+/// modules [TransportRpcModules] as well as configured authenticated methods [AuthRpcModule].
+#[allow(missing_debug_implementations)]
+pub struct RpcContext<'a, Node: FullNodeComponents> {
+    /// The node components.
+    pub(crate) node: Node,
+
+    pub(crate) config: &'a NodeConfig,
```

Comment:
guessing the full node config is here because it's just useful to have access to in methods like `on_rpc_started`? if so, docs for that would be good